### ISSUE TITLE
Implement search feature

### ISSUE THREAD

**Issue by justinyum98** [justinyum98/my-music-list-web#2]
**Is your feature request related to a problem? Please describe.**
The problem is that the user cannot search for items to review.

**Describe the solution you'd like**
As a user, I should be able to search for an item by providing a search query and the type of item to search for (artist, album, or track). Here is the proposed user flow:

1. User selects the item type to search for.
2. User types in search query.
3. User clicks a button / presses enter on keyboard to initiate search.
4. User receives search results.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by justinyum98** [comment]
@sangyum Do you think it's necessary to have another layer (e.g. music provider layer) separate from the service layer (i.e. music service)?
Within this music provider layer, the provider(s) (i.e. Spotify, Apple Music) will share the same responsibility: to provide music data to the music service. The only difference is where that music data is retrieved.

My reasoning for this is that the service layer, based on my assumptions, should be vendor-agnostic in that it shouldn't care where it gets the music data from.

So, to reiterate the main question: do you think it's necessary to add an extra layer? Or do you think that's premature optimization/abstraction?

**Comment by sangyum** [comment]
That is correct. It's called proxy pattern. https://refactoring.guru/design-patterns/proxy

If you put an abstraction layer on top of music provider (Spotify, Apple Music, YouTube Music, etc), the rest of your application can use a consistent interface to search music. Any changes in the music provider would be encapsulated in this layer. In fact, you can develop the rest of your application by mocking this interface.
